rust/abacus-core/src/traits/message.rs

```diff
@@ -21,7 +20,15 @@ impl RawCommittedMessage {
     /// The leaf is the keccak256 digest of the message, which is committed
     /// in the message tree
     pub fn leaf(&self) -> H256 {
-        keccak256(&self.message).into()
+        let buffer = [0u8; 28];
+        H256::from_slice(
+            Keccak256::new()
```

Comment:
I'm a little wary of doing any bespoke formatting of the contents here. Does the chaining here for sure always do the exact same thing as `abi.encodePacked`? The documentation on `abi.encodePacked` [here](https://docs.soliditylang.org/en/v0.8.13/abi-spec.html#non-standard-packed-mode) has me feeling like this is doing something different? If there's some documentation that explains the chaining here is correct it'd probably be good to leave it in a comment

To be fair I can't find anything in ethers-rs that does `abi.encodePacked`, just `abi.encode` from what I can see...

Reply:
What exactly do you mean by bespoke here? I just copied this from how we hash the digest for the checkpoints as well https://github.com/abacus-network/abacus-monorepo/blob/nambrot/multiple-inboxes-per-domain/rust/abacus-core/src/types/checkpoint.rs#L72:L72

While I don't think I have hard evidence, certainly the consistent use of this function makes me pretty confident that it works as intended. I think what might be confusing is the "buffer" of 28 0s. The buffer is not for ABI encoding, but the fact that on-chain, the leaf_index is a uint256, while in rust land it is a u32. Therefore we need to pad the leaf_index so that the pre-image is the same. Otherwise at least it seems to me that both `Keccak256::new.chain` and `abi.encodePacked`, basically just pack the bytes together

Reply:
ahhhh my bad, I didn't realize this was already present somewhere. By bespoke I meant circumventing any function that may exist already that matches abi.encodePacked behavior (like in ethers-rs or something). I'm ok with this implementation, thanks for the info